src/tasks/WasmAppBuilder/EmitWasmBundleFiles.cs

```diff
@@ -0,0 +1,268 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.Framework;
+
+namespace Microsoft.WebAssembly.Build.Tasks;
+
+public class EmitWasmBundleFiles : Microsoft.Build.Utilities.Task, ICancelableTask
+{
+    private CancellationTokenSource BuildTaskCancelled { get; } = new();
+
+    [Required]
+    public ITaskItem[] FilesToBundle { get; set; } = default!;
+
+    // if not provided the task will generate source files to disk
+    public string? ClangExecutable { get; set; }
```

Comment:
This is not specific to individual files nor to a type of bundle. It really only depends on compiler ability. The problem is that `emcc` can't receive code as input stream and that why we have 2 versions. I will add comment explaining that.

Reply:
Also I don't want to move it to different class. The difference between `WriteSource` and `Compile` is to large block of code.

Reply:
The size of code isn't the concern. My suggestion was because right now the behavior of the task changes magically depending on whether `ClangExecutable` was set or not.
1. one behavior is to emit `.c` source files for each of the files to be bundled
2. and the other is to further compile them, and emit the object files instead

The task should differentiate these two better. Maybe an additional parameter which opts-in to compiling.

Reply:
I split it into base class and 2 Tasks.